src/modem.rs

```diff
@@ -0,0 +1,259 @@
+use at_commands::{builder::CommandBuilder, parser::CommandParser};
+use atat::{self, AtatCmd};
+use core::{borrow::BorrowMut, ffi::c_void, marker::PhantomData};
+use esp_idf_hal::{delay::TickType, uart::UartDriver};
+
+use crate::{
+    handle::RawHandle,
+    netif::{EspNetif, NetifStack},
+    sys::*,
+};
+
+pub struct PppNetif<'d, T>
+where
+    T: BorrowMut<UartDriver<'d>>,
+{
+    serial: T,
+    base: esp_netif_driver_base_t,
+    netif: EspNetif,
+    _d: PhantomData<&'d ()>,
+}
+
+impl<'d, T> PppNetif<'d, T>
+where
+    T: BorrowMut<UartDriver<'d>>,
+{
+    pub fn new(serial: T) -> Result<Self, EspError> {
+        let netif = EspNetif::new(NetifStack::Ppp)?;
+        let mut base = esp_netif_driver_base_t {
+            netif: netif.handle(),
+            post_attach: Some(Self::post_attach),
+        };
+        let base_ptr: *mut c_void = &mut base as *mut _ as *mut c_void;
+        esp!(unsafe { esp_netif_attach(netif.handle(), base_ptr) })?;
+        Ok(Self {
+            serial,
+            netif,
+            base,
+            _d: PhantomData,
+        })
+    }
+
+    unsafe extern "C" fn post_attach(netif: *mut esp_netif_obj, args: *mut c_void) -> i32 {
+        let driver= unsafe{std::ptr::slice_from_raw_parts_mut(args, size_of::<ppp_netif_driver())}
+        let ifconfig = esp_netif_driver_ifconfig_t{handle}
```

Comment:
Really struggling to wrap my head around this part at the moment

Reply:
OK. Let me step back and try to explain `esp_netif_driver_base_t`, what is so base about it and how the ESP IDF folks envisioned it is going to be used.

It is called "base", as it is a form of "poor man's" inheritance in the C world, that - fortunately - can be used from within Rust as well.

* The ESP IDF assumes that you are going to have some sort of a PPP "driver" struct.  
* The closest in your current code to a driver is the `PppNetif` struct (but we won't use that directly, read on).
* When you call `esp_netif_attach`, you pass a `&mut` so to say, to your driver (say, to `PppNetif` or something similar)
* When the time had come for ESP-IDF to call `post_attach` (which would be much, much later, long after the call to `esp_netif_attach` is done and you have exited the `PppNetif::new` constructor), ESP IDF will do the following unsafe cast of the `void* base_ptr` that you pass to it: _it will try to cast it to `*esp_netif_base_t`_, so that it can extract from the `esp_netif_base_t` struct the `post_attach` callback so that it can call **you**
* "So far so good", you would say - "my `base_ptr` is anyway _A_ `esp_netif_base_t` instance?
  * And here we come to the "base" part. Yes, you are passing a pointer to a `esp_netif_base_t` instance so that would work, but the idea actually is to pass something _BIGGER_ that `esp_idf_base_t` to `esp_netif_attach`, which contains - besides `esp_netif_base_t` - _also your own "user" data_ - user data, that _you_ can "get back" when your own `PppNetif::post_attach` is called and that's data _that you might need_ so that _your_ `PppNetif::post_attach` can do its job.

The following magical struct would work and you can pass to ESP IDF as a `base_ptr`  (I mean, you can pass a _pointer_ to it, of course, not the "whole" struct):
```rust
#[repr(C)]
struct MyDriver {
    base: esp_netif_base_t,
    netif: EspNetif,
    foo: Foo,
    bar: Bar,
    ... more stuff
}
```

Since it is `repr(C)` (i.e. fields-reordering that Rust sometimes does is forbidden) and since `base` is the FIRST member of this struct, when ESP IDF tries to cast `&MyDriver` to `&esp_netif_base_t` it would actually work! And the benefit is that you are passing to ESP IDF something much larger. Still - ESP IDF is able to cast it to something smaller (to esp_netif_base_t, because, well, it is the _first_ member in the struct layout), extract from the "smaller" part what it needs (the netif handle and your `post_attach` callback) and then - _when the time had come_ - call your `post_attach` callback with a pointer to `MyDriver`, so that you have your own data in your `post_attach` function.

OK, seemingly all is well and good, and you might be tempted to slap a `#[repr(C)]` on top of your `PppNetif` struct and then pass a `&PppNetif` reference to `esp_netif_attach`  (strictly speaking a `&PppNetif as *const _ as *const c_void as *mut _`).

Will that work?
No.

...which is the second issue you need to workaround: unlike C (and C++), Rust has _move_semantics. What it means, is that unless you take special care, passing a `&some_rust_struct` might crash, because at the time when the C code needs to use `&some_rust_struct`, `some_rust_struct` might have _moved_ to another memory location, so `&some_rust_struct` - at that time - is **invalid**.

In your particular case, passing `&PppNetif` to `esp_netif_attach` - _from within `PppNetif::new`_ - is not OK, because - even if you construct `PppNetif` eariler in `new` - after `new` is done, `PppNetif` will be _moved_ to a new location (as `new` returns `Self` = `PppNetif`)...

The easiest way to workaround this is to pass to `esp_netif_attach` something else which:
* Still contains all the data you'll need in `PppNetif::post_attach`
* Is **boxed**. I.e. if we call this type `PppNetifDriver`, you need to - in `PppNetif::new` - create it with `let driver = Box::new(PppNetifDriver { ... })` and then pass to `esp_netif_attach` a `driver.as_ptr()` pointer to the box
* ... and of course you are free to preserve `driver` (`Box<PppNetifDriver>`) inside `PppNetif` so that at the very least, the heap memory is dropped when the user drops `PppNetif` and also the destructor of `PppNetifDriver` (if any) runs as part of the `PppNetif` destructor
* **The most important is**: since the box is allocated on-heap, when `PppNetif` is moved around, `PppNetifDriver` **won't move**. It is on the heap after all. So you can pass a reference to the heap memory to the ESP IDF C code, and (as long as `PppNetif` is still alive), that heap memory would be valid and OK to use.
* Obviously, in `PppNetif::post_attach` you need to do `unsafe { (args as *mut PppNetifDriver).as_mut() }.unwrap()` and you'll get a `&mut PppNetifDriver` and then you can use it however you want to finish the "post-attached" logic.

(
Digression:
The above "anti-move" approach utilizes a heap allocation. While this is sub-optimal:
- ESP IDF itself allocates like mad, so I think we doing an allocation in these circumstances (one-time creation of a long-living object - the driver) is justified
- The alternative (Rust pinned types) is considered a complex animal in Rust, and it will _affect the API of `PppNetif`_ in that (a) initialization will become two-phase; (b) users would be required to pin `PppNetif` before executing the second-phase initialization. Given that there is a large base of users who reach out to the ESP IDF wrappers not because they are more complete, but because they are perceived as "simpler" than bare-metal, we definitely do not want to introduce pinning in the API.


)

Hope that helps. Please follow up with more questions if something is unclear.

Reply:
Thinking about your code, I feel the easiest way to apply the patterns I'm talking about might be to do something like:
```rust
pub struct PppNetifDriver {
    netif: EspNertif,
    inner: Box<PppNetifDriverInner>,
}

impl PppNetifDriver {
    /// Create a new PPP driver with the provided ipv4 configuration and a TX
    /// callback that the driver would call when it wants to ingest a PPP packet into the 
    /// underlying transport.
    /// The transport can be anything, but would typically be UART, and the callback implementation
    /// is simply expected to write the PPP packet into UART
    pub fn new<F>(conf: &crate::ipv4::Configuration, tx: F) -> Result<Self, EspError> 
    where 
        F: FnMut(&[u8]) -> Result<(), EspError>,
    {
        let netif = EspNetif::new(...);
        let inner = Box::new(PppNetifDriverInner {
            base: esp_netif_driver_base_t {
                netif: netif.handle(),
                post_attach: PppNetifDriverInner::raw_post_attach,
            },
            tx: Box::new(tx),
        });

        let inner_ptr = inner.as_ptr() as *const c_void as *mut _;

        esp!(unsafe { esp_event_handler_register(NETIF_PPP_STATUS, ESP_EVENT_ANY_ID, netif.handle()) })?;
        esp!(unsafe { esp_event_handler_register(IP_EVENT, IP_EVENT_PPP_GOT_IP, esp_netif_action_connected, netif.handle()) })?;
        esp!(unsafe { esp_event_handler_register(IP_EVENT, IP_EVENT_PPP_LOST_IP, esp_netif_action_disconnected, netif.handle()) })?;

        esp!(unsafe { esp_netif_attach(netif.handle(), inner_ptr) })?;

        Ok(Self {
            netif,
            inner,
        })
    }

    /// Ingest a PPP packet into the driver. The PPP packet can arrive from anywhere, but 
    /// typically from UART, by reading from it.
    pub fn rx(&self, data: &[u8]) -> Result<(), EspError> {
        self.inner.rx()
    }
}

#[repr(C)]
struct PppNetifDriverInner {
    base: esp_netif_driver_base_t,
    tx: Box<dyn FnMut(&[u8]) -> EspError>, // Boxing only necessary so as to erase `F` from the driver type signature; if we keep it, then we can geit rid of this particular boxing. But given that we box anyway the inner driver, what gives?
}

impl PppNetifDriverInner {
    fn post_attach(&mut self, netif_handle: *mut esp_netif_obj) -> Result<(), EspError> {
    }

    unsafe extern "C" fn post_attach(netif: *mut esp_netif_obj, args: *mut c_void) -> i32 {
        let this = (args as *mut PppNetifDriverInner).as_mut() }.unwrap();
        match this.post_attach(netif) {
            Ok(_) => ESP_OK,
            err(e) => e,code(),
        }
    }
}
```

Reply:
Heh.
Now that I think about that code from above some more, _there is nothing PPP-specific in that code_!
It is just a way to attach an RX/TX functionality to a Netif so that you ingest/ingress (RX) packets into the netif, and you can egress (TX) packets from the netif. It might be used with a PPP Netif, or it might be used with - whatever.

So if we fix the naming and generalize the code just a tad more, we end up with:
```rust
pub struct EspNetifDriver<T> 
where 
	T: Borrow<EspNetif>,
{
    inner: Box<EspNetifDriverInner<T>>,
}

impl<T> EspNetifDriver<T> 
where 
	T: Borrow<EspNetif>,
{
    /// Create a new netif driver around the provided `EspNetif` instance and a TX
    /// callback that the driver would call when it wants to ingest a packet into the 
    /// underlying transport.
    /// 
    /// The transport can be anything, but with - say - PPP netif - it would typically be UART, and the callback implementation
    /// is simply expected to write the PPP packet into UART
    ///
    /// `post_attach_cfg` is a netif-specifif configuration that will be executed after the netif is attached
    /// See `ppp_config` below for a PPP-specific driver post-attach configuration
    pub fn new<F, P>(netif: T, post_attach_cfg: P, tx: F) -> Result<Self, EspError> 
    where 
        F: FnMut(&[u8]) -> Result<(), EspError>,
        P: FnMut(&EspNetif) -> Result<(), Error>,
    {
        let inner = Box::new(PppNetifDriverInner {
            base: esp_netif_driver_base_t {
                netif: netif.borrow().handle(),
                post_attach: PppNetifDriverInner::raw_post_attach,
            },
            netif,
            post_attach_cfg: Box::new(post_attach_cfg),
            tx: Box::new(tx),
        });

        let inner_ptr = inner.as_ptr() as *const c_void as *mut _;

        // TODO: We should rather use the safe event loop API
        // TODO: May not even be necessary to do this from _within_ the driver - it can be done completely in user code
        // TODO Also, PPP-specific, and best left to the user to do, from outside
        // esp!(unsafe { esp_event_handler_register(NETIF_PPP_STATUS, ESP_EVENT_ANY_ID, PppNetifDriverInner::raw_on_ppp_changed) })?;

        // TODO: Check if `EspNetif` is not doing this already...
        esp!(unsafe { esp_event_handler_register(IP_EVENT, IP_EVENT_PPP_GOT_IP, esp_netif_action_connected, inner.netif.borrow().handle()) })?;
        // TODO: Ditto
        esp!(unsafe { esp_event_handler_register(IP_EVENT, IP_EVENT_PPP_LOST_IP, esp_netif_action_disconnected, inner.netif.borrow().handle()) })?;

        esp!(unsafe { esp_netif_attach(inner.netif.borrow().handle(), inner_ptr) })?;
        esp!(unsafe { esp_netif_action_start(inner.netif.borrow().handle(), core::ptr::null_mut(), 0, core::ptr::null_mut()) })?;

        Ok(Self {
            inner,
        })
    }

    /// Ingest a packet into the driver. 
    /// The packet can arrive from anywhere, but with say - a PPP netif - it would be a PPP packet arriving typically from UART, by reading from it.
    pub fn rx(&self, data: &[u8]) -> Result<(), EspError> {
        esp!(unsafe { esp_netif_receive(self.inner.netif.borrow().handle(), data.as_raw_ptr() as *mut c_void, len as _, core::ptr::null_mut()) })
    }
}

impl<T> Drop for EspNetifDriver<T> 
where 
	T: Borrow<EspNetif>,
{
    fn drop(&mut self) {
        esp!(unsafe { esp_netif_action_stop(self.inner.netif.borrow().handle(), core::ptr::null_mut(), 0, core::ptr::null_mut()) }).unwrap();

        // TODO: See constructor
        // esp!(unsafe { esp_event_handler_unregister(NETIF_PPP_STATUS, ESP_EVENT_ANY_ID, &EspNetifDriverInner::raw_on_ppp_changed) }).unwrap();

        // TODO: See constructor
        esp!(unsafe { esp_event_handler_unregister(IP_EVENT, IP_EVENT_PPP_GOT_IP, esp_netif_action_connected) }).unwrap();
        // TODO: See constructor
        esp!(unsafe { esp_event_handler_unregister(IP_EVENT, IP_EVENT_PPP_LOST_IP, esp_netif_action_disconnected) }).unwrap();
    }
}

#[repr(C)]
struct EspNetifDriverInner<T> 
where 
	T: Borrow<EspNetif>,
{
    base: esp_netif_driver_base_t,
    tx: Box<dyn FnMut(&[u8]) -> EspError>, // Boxing only necessary so as to erase `F` from the driver type signature; if we keep it, then we can geit rid of this particular boxing. But given that we box anyway the inner driver, what gives?
    netif: T,
}

impl<T> EspNetifDriverInner<T> 
where 
	T: Borrow<EspNetif>,
{
    fn post_attach(&mut self, netif_handle: *mut esp_netif_obj) -> Result<(), EspError> {
        let driver_ifconfig = esp_netif_driver_ifconfig_t {
            transmit = Self::raw_tx;
            handle = self as *mut c_void,
        };

        // d->base.netif = esp_netif; TODO: This is weird; the netif in base is already set on constructor?
        
        esp!(unsafe { esp_netif_set_driver_config(netif_handle, &driver_ifconfig) })?;

		(self.post_attach_cfg)()?;

        Ok(())
    }

    fn tx(&mut self, data: &[u8]) -> Result<(), EspError> {
        (self.tx)(data)
    }

    fn on_ppp_changed(&mut self, event_base: esp_event_base_t,
                           event_id: i32, event_data: *const c_void) {
        if event_id > NETIF_PPP_ERRORNONE && event_id < NETIF_PP_PHASE_OFFSET {
            todo!()
            // ESP_LOGI("esp_modem_netif", "PPP state changed event %" PRId32, event_id);
            // only notify the modem on state/error events, ignoring phase transitions
            // ppp->signal.set(PPP_EXIT);
        }
    }

    unsafe extern "C" fn raw_tx(h: *mut c_void, buffer: *mut c_void, len: usize) -> i32 {
        let this = (args as *mut Self).as_mut() }.unwrap();
        let data = core::slice::from_raw_parts(buffer as *mut u8, len);
        let result = match this.tx(data) {
            Ok(_) => ESP_OK,
            err(e) => e,code(),
        };

        // TODO: Might not be necessary, but if I remember correctly, the Netif API
        // wanted that _we_ free the buffer; in any case needs to be compared with the C ESP Modem code
        free(buffer);

        result
    }

    unsafe extern "C" fn raw_post_attach(netif: *mut esp_netif_obj, args: *mut c_void) -> i32 {
        let this = (args as *mut Self).as_mut() }.unwrap();
        match this.post_attach(netif) {
            Ok(_) => ESP_OK,
            err(e) => e,code(),
        }
    }

    unsafe extern "C" fn raw_on_ppp_changed(arg: *mut c_void, event_base: esp_event_base_t,
                           event_id: i32, event_data: *const c_void) {
        let this = (args as *mut Self).as_mut() }.unwrap();
        this.on_ppp_changed(event_base, event_id, event_data);        
    }
}

/// PPP-specific configuration of a Netif
pub fn ppp_config(netif: &EspNetif) -> Result<(), EspError> {
    // check if PPP error events are enabled, if not, do enable the error occurred/state changed
    // to notify the modem layer when switching modes
    let mut ppp_config = esp_netif_ppp_config_t {
        ppp_phase_event_enabled = true,    // assuming phase enabled, as earlier IDFs
        ppp_error_event_enabled = false,   // don't provide cfg getters so we enable both events
        ...Default::default()
    };
    
    #[cfg(not(esp_idf_version_major = "4"))]
    esp!(unsafe { esp_netif_ppp_get_params(esp_netif, &ppp_config) })?;

    if !ppp_config.ppp_error_event_enabled {
        ppp_config.ppp_error_event_enabled = true;
        esp!(unsafe { esp_netif_ppp_set_params(esp_netif, &ppp_config) })?;
    }
    
    Ok(())
}
```

Reply:
> Since it is repr(C) (i.e. fields-reordering that Rust sometimes does is forbidden) and since base is the FIRST member of this struct, when ESP IDF tries to cast &MyDriver to &esp_netif_driver_base_t it would actually work! And the benefit is that you are passing to ESP IDF something much larger. Still - ESP IDF is able to cast it to something smaller (to esp_netif_driver_base_t, because, well, it is the first member in the struct layout), extract from the "smaller" part what it needs (the netif handle and your post_attach callback) and then - when the time had come - call your post_attach callback with a pointer to MyDriver, so that you have your own data in your post_attach function.

The fact that casting `EspNetifDriverInner` to   `esp_netif_driver_base_t` takes the first member feels like arcane magic. Brilliant.